### PR TITLE
Add subcommands `plugins {list,status}`, to inquire plugins

### PR DESCRIPTION
**Pull request by amotl**
## About

At https://github.com/panodata/grafana-client/pull/110, @bhks added a few more API wrapper functions for Grafana. Thanks! This patch wraps them once more into the command line interface of grafana-wtf.

## Synopsis
```
# Explore plugins.
grafana-wtf plugins list
grafana-wtf plugins status
```

### CODE REVIEW

**Comment by amotl** [comment]
@bhks: Even with the most recent Grafana 10.1.2 release, `grafana-wtf plugins status` does not work well on my machine. It fails to inquire the corresponding health check and metrics endpoints.

- Maybe I am using them wrong?
- Do they need to be enabled within `grafana.ini`?
- On the other hand, maybe both are enterprise features, and not available in Grafana OSS?

You can install the package including this feature directly from the corresponding branch using this `pip` command, in order to check if it works on your end. I will be happy to hear back about the outcome.
```shell
pip install --upgrade 'git+https://github.com/panodata/grafana-wtf@list-plugins'
```

**Comment by bhks** [comment]
Thanks for pinging me into this thread.

The panel plugin does not have any server running which can give us health or metrics from the backend. Also not all of the plugin servers implement this.

The After plugin sdk kind of enforces for plugin developers to implement the protobuff model from version-1, so I don't know if all of the marketplace plugins of type datasource or App follows those.

**Comment by bhks** [comment]
I can try this out when I get a chance and see what works and what not.

File: grafana_wtf/core.py [hunk 1]
```diff
@@ -492,6 +492,36 @@ def explore_dashboards(self, with_data_details: bool = False):
...
 
+    def plugins_list(self):
+        return self.grafana.plugin.get_installed_plugins()
+
+    def plugins_status(self):
+        status = []
+        plugins = self.grafana.plugin.get_installed_plugins()
+        for plugin in plugins:
+            plugin = munchify(plugin)
+            item = Munch(
```
Comment (bhks) [inline grafana_wtf/core.py:503]: I think we can filter out the panel type as they don't have server to respond from.

**Comment by bhks** [comment]
Not all servers/plugins have implemented the metric endpoint, this was I believe introduced in the new SDK which helps build an external plugin.

Also all core plugins does not have a server so they are built within grafana golang code , they don't have health or metric endpoint. So we need to filter them out as well.

**Comment by amotl** [comment]
Thanks for your response. According to `grafana-wtf plugins list`, a representation looks like this.

```json
{
    "name": "Alert list",
    "type": "panel",
    "id": "alertlist",
    "enabled": true,
    "pinned": false,
    "info": {
        "author": {
            "name": "Grafana Labs",
            "url": "https://grafana.com"
        },
        "description": "Shows list of alerts and their current status",
        "links": null,
        "logos": {
            "small": "public/app/plugins/panel/alertlist/img/icn-singlestat-panel.svg",
            "large": "public/app/plugins/panel/alertlist/img/icn-singlestat-panel.svg"
        },
        "build": {},
        "screenshots": null,
        "version": "",
        "updated": ""
    },
    "dependencies": {
        "grafanaDependency": "",
        "grafanaVersion": "*",
        "plugins": []
    },
    "latestVersion": "",
    "hasUpdate": false,
    "defaultNavUrl": "/grafana",
    "category": "",
    "state": "",
    "signature": "internal",
    "signatureType": "",
    "signatureOrg": ""
}
```

> Also all core plugins does not have a server so they are built within grafana golang code , they don't have health or metric endpoint. So we need to filter them out as well.

Would skipping all plugins having `"signature": "internal"` on health and metrics inquiry a good option to proceed with here?

**Comment by bhks** [comment]
>Would skipping all plugins having "signature": "internal" on health and metrics inquiry a good option to proceed with here?

Exactly that and the following one as well

```
"type": "panel",
```

**Comment by amotl** [comment]
Maybe `"type": "datasource"` would be the right choice, regardless of `"signature": "internal"` and `"type": "panel"` or not?

**Comment by amotl** [comment]
I've amended the patch to only use `if item.type == "datasource"` at this spot, and I think it works well so far.

**Comment by bhks** [comment]
But internal/core plugins will not be able to respond like Cloudwatch, prometheus, rds they are datasource type.

**Comment by amotl** [comment]
Thanks. I've added `if item.type == "datasource" and item.signature != "internal"` to my working tree, and it significantly reduces unneccessary probes further.

**Comment by bhks** [comment]
Do you mind updating that if with following :

```
if item.type != "panel" and item.signature != "internal"
```

File: tests/test_commands.py [hunk 2]
```diff
@@ -555,3 +556,60 @@ def test_info(docker_grafana, capsys, caplog):
...
+        grafana_wtf.commands.run()
+    captured = capsys.readouterr()
+    data = yaml.safe_load(captured.out)
+
+    # Grafana 6 has 28 plugins preinstalled.
+    assert len(data) >= 28
+
+    # Proof the output is correct.
+    infinity_datasource = next(item for item in data if item["id"] == "yesoreyeram-infinity-datasource")
+    assert "go_gc_duration_seconds" in infinity_datasource["metrics"]
```
Comment (amotl) [inline tests/test_commands.py:615]: @bhks: This is a first real integration test, using your contribution to the grafana-client library.

Comment (bhks) [reply tests/test_commands.py:615]: Awesome work Andreas.

**Comment by amotl** [comment]
> Do you mind updating that if with following :
>
> ```
> if item.type != "panel" and item.signature != "internal"
> ```

Can you convince me why this is better? To me, it sounds more inadequate, because we are mostly talking about "datasource" plugins here? Are there any other types of plugins which yield sensible responses on their metrics or health endpoints?

If so, can you provide a sample (plugin id) of that kind, so I can use it on behalf of a corresponding software test case? Thanks!

**Review by bhks — APPROVED** [review]

File: tests/test_commands.py (hunk 2 above)
Comment (amotl) [reply tests/test_commands.py:615]: Just building upon _your_ work here ;], thanks again.

Comment (amotl) [reply tests/test_commands.py:615]: CI currently fails here, because a few more adjustments will be submitted to grafana-client to make it work.

**Comment by bhks** [comment]
> Can you convince me why this is better? To me, it sounds more inadequate, because we are mostly talking about "datasource" plugins here? Are there any other types of plugins which yield sensible responses on their metrics or health endpoints?
> 
> If so, can you provide a sample (plugin id) of that kind, so I can use it on behalf of a corresponding software test case?

Apologies about being too specific here.

May be the example plugin id in comment can help :https://github.com/panodata/grafana-wtf/pull/91#issuecomment-1728543881

I was referring to a third type which is App plugin type which have data source as well as panel plugins bundled. Example like 

`aws-datasource-provisioner-app`.

Here are multiple app type plugins we can explore , I am not 100% sure if they all have these endpoints : https://grafana.com/grafana/plugins/app-plugins/

File: tests/test_commands.py (hunk 2 above)
Comment (amotl) [reply tests/test_commands.py:615]: Also:

> TODO: Verify a plugin which properly responds to a **health check** response.

Haven't been able to discover one, yet. If that plugin sample you just shared with me, will respond to such a response, I will be happy about it.

**Review by bhks — APPROVED** [review]

**Comment by amotl** [comment]
`aws-datasource-provisioner-app` works well, and provides both health and metrics. Thank you!
[
    {
        "name": "AWS Data Sources",
        "type": "app",
        "id": "aws-datasource-provisioner-app",
        "enabled": false,
        "category": "",
        "version": "1.13.0",
        "signature": "valid",
        "health": {
            "message": "",
            "status": "OK"
        },
        "metrics": "# HELP go_cgo_go_to_c_calls_calls_total Count of calls made from Go to C by the current process.\n# TYPE go_cgo_go_to_c_calls_calls_total counter\ngo_cgo_go_to_c_calls_calls_total 0\n# HELP go_cpu_classes_gc_mark_assist_cpu_seconds_total Estimated total CPU time goroutines spent performing GC tasks to assist the GC and prevent it from falling behind the application. This metric is an overestimate, and not directly

**Comment by amotl** [comment]
Needs https://github.com/panodata/grafana-client/pull/115.

File: grafana_wtf/core.py [hunk 3]
```diff
@@ -492,6 +492,40 @@ def explore_dashboards(self, with_data_details: bool = False):
...
```
Comment (bhks) [inline grafana_wtf/core.py:503]: Wow I haven't used `munchify` ever , that seems like a great way to remove the tedious key value mapping with strings.

File: grafana_wtf/core.py [hunk 4]
```diff
@@ -492,6 +492,40 @@ def explore_dashboards(self, with_data_details: bool = False):
...
+                id=plugin.id,
+                enabled=plugin.enabled,
+                category=plugin.category,
+                version=plugin.info.version,
+                signature=plugin.get("signature"),
+            )
+
+            # Status inquiry is not provided by all plugins. Let's filter them.
+            # Effectively, run it only on non-internal "app" and "datasource" items.
+            if item.type != "panel" and item.signature != "internal":
```
Comment (bhks) [inline grafana_wtf/core.py:515]: Thank you for incorporating !!

File: grafana_wtf/core.py [hunk 5]
```diff
@@ -492,6 +492,40 @@ def explore_dashboards(self, with_data_details: bool = False):
...
+                    item.health = self.grafana.plugin.health_check_plugin(plugin.id)
+                except Exception as ex:
+                    log.warning(f"Health check failed for plugin {item.id}, type={item.type}: {ex}")
+                try:
+                    item.metrics = self.grafana.plugin.get_plugin_metrics(plugin.id)
+                except Exception as ex:
+                    log.warning(f"Metrics inquiry failed for plugin {item.id}, type={item.type}: {ex}")
+            else:
+                log.info(f"Skipping status inquiry for plugin {item.id}, type={item.type}")
+            status.append(item)
```
Comment (bhks) [inline grafana_wtf/core.py:526]: Nice in one go we can get all of them.

**Review by bhks — APPROVED** [review]

File: grafana_wtf/core.py (hunk 3 above)
Comment (amotl) [reply grafana_wtf/core.py]: If you can afford the runtime speed bump, it definitively saves a few keystrokes!

By the way, big thanks to @dsc, @vmalloc, @JosePVB, @innermatrix, @d1618033, and all others who conceived it, and are maintaining it.

**Pull request merged**